src/Tools/dotnet-monitor/Egress/S3/S3StorageEgressProvider.cs

```diff
@@ -0,0 +1,258 @@
+                await action(stream, token);
+                await stream.FinalizeAsync(token); // force to push the last part
+                stream.Close();
+
+                if (stream.Parts.Count == 0)
+                {
+                    Logger?.LogDebug(Strings.Message_EgressS3NoContent);
+                    return string.Empty; // nothing to return here
+                }
+
+                await CompleteMultiPartUploadAsync(client, options.BucketName, uploadId, artifactSettings, stream.Parts, token);
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        private async Task<string> UploadToStorageAsync(
+            S3StorageEgressProviderOptions options,
+            Stream stream,
+            EgressArtifactSettings artifactSettings,
+            CancellationToken token)
+        {
+            try
+            {
+                var client = await CreateClientAsync(options, token);
+                if (stream.CanSeek) // use the stream directly
+                {
+                    var request = new PutObjectRequest
+                    {
+                        BucketName = options.BucketName,
+                        ContentType = artifactSettings.ContentType,
+                        Key = artifactSettings.Name,
+                        InputStream = stream,
+                        AutoCloseStream = false,
+                    };
+                    await client.PutObjectAsync(request, token);
+                }
+                else // copy temporary to memory stream locally
+                {
+                    using TransferUtility transferUtility = new(client);
+                    await transferUtility.UploadAsync(stream, options.BucketName, artifactSettings.Name, token);
+                }
+
+                string resourceId = GetResourceId(client, options, artifactSettings);
+                return resourceId;
+            }
+            catch (Exception e)
+            {
+                throw CreateException(string.Format(CultureInfo.CurrentCulture, Strings.ErrorMessage_EgressS3FailedDetailed, e.Message));
+            }
+        }
+
+        private string GetResourceId(IAmazonS3 client, S3StorageEgressProviderOptions options, EgressArtifactSettings artifactSettings)
+        {
+            if (!options.GeneratePresSignedUrl)
+                return $"BucketName={options.BucketName}, Key={artifactSettings.Name}";
+
+            DateTime expires = DateTime.UtcNow.AddMinutes(options.PreSignedUrlExpiryInMinutes);
+            string resourceId = client.GetPreSignedURL(new GetPreSignedUrlRequest { BucketName = options.BucketName, Key = artifactSettings.Name, Expires = expires });
```

Comment:
This looks like it produces a URL that allows access to the resource for some time. Will this always succeed given all of the ways that credentials seem to be given for this egress provider? It should not fail the egress operation just because it cannot get this URL.

In the Azure blob egress provider, we've elected to only return a URL that represents the path to the resource without any signed token that allows access to avoid issues like this as well as to avoid giving access to resource directly through dotnet-monitor.

Consider just returning a URL that is the path to the resource without access token information.